File: Algorithms/src/865. Smallest Subtree with all the Deepest Nodes/Code/solution.go
func subtreeWithAllDeepest(root *TreeNode) *TreeNode {

    var dfs func(*TreeNode) (int, *TreeNode)

    dfs = func(node *TreeNode) (int, *TreeNode) {
        if node == nil {
            return 0, nil
        }

        ld, ln := dfs(node.Left)
        rd, rn := dfs(node.Right)

        if ld == rd {
            return ld + 1, node
        } else if ld > rd {
            return ld + 1, ln
        } else {
            return rd + 1, rn
        }
    }

    _, ans := dfs(root)
    return ans
}

File: Algorithms/src/865. Smallest Subtree with all the Deepest Nodes/Code/solution.js
var subtreeWithAllDeepest = function (root) {
  function dfs(node) {
    if (!node) return [0, null];

    const [ld, ln] = dfs(node.left);
    const [rd, rn] = dfs(node.right);

    if (ld === rd) {
      return [ld + 1, node];
    } else if (ld > rd) {
      return [ld + 1, ln];
    } else {
      return [rd + 1, rn];
    }
  }

  return dfs(root)[1];
};
